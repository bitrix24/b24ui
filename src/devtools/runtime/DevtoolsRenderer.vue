<script setup lang="ts">
import { onUnmounted, onMounted, reactive } from 'vue'
import { pascalCase } from 'scule'
import { defineAsyncComponent, useColorMode, useRoute } from '#imports'

const route = useRoute()
const component = route.query?.example
  ? defineAsyncComponent(() => import(`./examples/${route.query.example}.vue`))
  : route.params?.slug && defineAsyncComponent(() => import(`../../runtime/components/${pascalCase(route.params.slug as string)}.vue`))

const state = reactive<{ slots?: any, props?: any }>({})

function onUpdateRenderer(event: Event & { data?: any }) {
  state.props = { ...event.data.props }
  state.slots = { ...event.data.slots }
}

const colorMode = useColorMode()
function setColorMode(event: Event & { isDark?: boolean }) {
  colorMode.preference = event.isDark ? 'dark' : 'light'
}

onMounted(() => {
  window.parent.addEventListener('bitrix24-ui-devtools:update-renderer', onUpdateRenderer)
  window.parent.addEventListener('bitrix24-ui-devtools:set-color-mode', setColorMode)
})

onUnmounted(() => {
  window.parent.removeEventListener('bitrix24-ui-devtools:update-renderer', onUpdateRenderer)
  window.parent.removeEventListener('bitrix24-ui-devtools:set-color-mode', setColorMode)
})

onMounted(async () => {
  const event: Event = new Event('bitrix24-ui-devtools:component-loaded')
  window.parent.dispatchEvent(event)
})

onMounted(() => {
  if (!route.query?.example) return
})
</script>

<template>
  <div id="b24ui-devtools-renderer" class="bitrix24-ui-component-renderer">
    <B24App :toaster="null">
      <component :is="component" v-if="component" v-bind="state.props" :class="state?.slots?.base" :ui="state.slots" />
    </B24App>
  </div>
</template>

<style>
.bitrix24-ui-component-renderer {
  position: 'relative';
  height: 100vh;
  width: 100vw;

  padding: 32px;

  display: flex;
  justify-content: center;
  align-items: center;

  background-image: url("data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http://www.w3.org/2000/svg' width='20' height='20' transform='scale(3)'%3E%3Crect width='100%25' height='100%25' fill='%23fff'/%3E%3Cpath fill='none' stroke='hsla(0, 0%25, 98%25, 1)' stroke-width='.2' d='M10 0v20ZM0 10h20Z'/%3E%3C/svg%3E");
  background-size: 40px 40px;
}

.dark .bitrix24-ui-component-renderer {
background-image: url("data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http://www.w3.org/2000/svg' width='20' height='20' transform='scale(3)'%3E%3Crect width='100%25' height='100%25' fill='hsl(0, 0%25, 8.5%25)'/%3E%3Cpath fill='none' stroke='hsl(0, 0%25, 11.0%25)' stroke-width='.2' d='M10 0v20ZM0 10h20Z'/%3E%3C/svg%3E");
  background-size: 40px 40px;
}
</style>
