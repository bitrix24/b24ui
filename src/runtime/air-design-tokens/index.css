@import './001_b24_global.css';
@import './003_b24_context-content-light.css';
@import './004_b24_context-content-dark.css';
@import './005_b24_context-edge-light.css';
@import './006_b24_context-edge-dark.css';
@import './007_b24_global.css';
@import './008_ui_global.css';
@import './009_b24_tools.css';
@import './components/scrollbar.css';
@import './components/navigation-menu.css';
@import './components/badge-counter.css';
