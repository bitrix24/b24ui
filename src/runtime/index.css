@plugin '@bitrix24/b24style';
@import '#build/b24ui.css';
@import './keyframes.css';
@import './air-design-tokens/index.css';

@layer base {
  body,
  .sidebar-layout.--inner {
    /* @see air-design-tokens/???_b24_context*.css */
    --air-theme-background:
      var(--air-theme-bg-image, none)
      var(--air-theme-bg-position, 0 0) / var(--air-theme-bg-size, auto)
      var(--air-theme-bg-repeat, repeat)
      var(--air-theme-bg-attachment, scroll)
      var(--air-theme-bg-color, transparent)
    ;

    --air-theme-background-blurred:
      var(--air-theme-bg-image-blurred, none)
      var(--air-theme-bg-position, 0 0) / var(--air-theme-bg-size, auto)
      var(--air-theme-bg-repeat, repeat)
      var(--air-theme-bg-attachment, scroll)
    ;
  }

  /* @see src/theme/sidebar-layout.ts */
  body {
    background: var(--air-theme-background);
  }

  .sidebar-layout.--inner {
    background: var(--air-theme-background);
  }

  body {
    overflow-y: scroll!important;
    padding-right: 0!important;
    scrollbar-gutter: stable;
  }

  .--app[data-state="loading"] {
    overflow: hidden;
  }
}
