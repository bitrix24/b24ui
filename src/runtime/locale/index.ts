/**
 * en English
 * de Deutsch
 * la Español
 *
 * br Português (Brasil)
 * fr Français
 * it Italiano
 *
 * pl Polski
 * ru Русский
 * ua Українська
 *
 * tr Türkçe
 * sc 中文（简体）
 * tc 中文（繁體)
 *
 * ja 日本語
 * vn Tiếng Việt
 * id Bahasa Indonesia
 *
 * ms Bahasa Melayu
 * th ภาษาไทย
 * ar عربي,
 *
 * kz Қазақша
 */
export { default as en } from './en'
export { default as de } from './de'
export { default as la } from './la'

export { default as br } from './br'
export { default as fr } from './fr'
export { default as it } from './it'

export { default as pl } from './pl'
export { default as ru } from './ru'
export { default as ua } from './ua'

export { default as tr } from './tr'
export { default as sc } from './sc'
export { default as tc } from './tc'

export { default as ja } from './ja'
export { default as vn } from './vn'
export { default as id } from './id'

export { default as ms } from './ms'
export { default as th } from './th'
export { default as ar } from './ar'

export { default as kz } from './kz'
