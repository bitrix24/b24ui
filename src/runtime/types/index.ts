export * from '../components/App.vue'
export * from '../components/Avatar.vue'
export * from '../components/AvatarGroup.vue'
export * from '../components/Button.vue'
export * from '../components/Checkbox.vue'
export * from '../components/Chip.vue'
export * from '../components/Container.vue'
export * from '../components/FormField.vue'
export * from '../components/Input.vue'
export * from '../components/Kbd.vue'
export * from '../components/Link.vue'
export * from '../components/Progress.vue'
export * from '../components/RadioGroup.vue'
export * from '../components/Separator.vue'
export * from '../components/Skeleton.vue'
export * from '../components/content/SkeletonTaskAdd.vue'
export * from '../components/Switch.vue'
export * from '../components/Textarea.vue'
export * from '../components/Tooltip.vue'
export * from './form'
export * from './locale'
