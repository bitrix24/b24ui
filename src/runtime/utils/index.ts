import { isEqual } from 'ohash/utils'
import type { GetItemKeys, NestedItem } from '../types/utils'

export function pick<Data extends object, Keys extends keyof Data>(data: Data, keys: Keys[]): Pick<Data, Keys> {
  const result = {} as Pick<Data, Keys>

  for (const key of keys) {
    result[key] = data[key]
  }

  return result
}

export function omit<Data extends object, Keys extends keyof Data>(data: Data, keys: Keys[]): Omit<Data, Keys> {
  const result = { ...data }

  for (const key of keys) {
    // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
    delete result[key]
  }

  return result as Omit<Data, Keys>
}

export function get(object: Record<string, any> | undefined, path: (string | number)[] | string, defaultValue?: any): any {
  if (typeof path === 'string') {
    path = path.split('.').map((key) => {
      const numKey = Number(key)
      return Number.isNaN(numKey) ? key : numKey
    })
  }

  let result: any = object

  for (const key of path) {
    if (result === undefined || result === null) {
      return defaultValue
    }

    result = result[key]
  }

  return result !== undefined ? result : defaultValue
}

export function set(object: Record<string, any>, path: (string | number)[] | string, value: any): void {
  if (typeof path === 'string') {
    path = path.split('.').map((key) => {
      const numKey = Number(key)
      return Number.isNaN(numKey) ? key : numKey
    })
  }

  path.reduce((acc, key, i) => {
    if (acc[key] === undefined) acc[key] = {}
    if (i === path.length - 1) acc[key] = value
    return acc[key]
  }, object)
}

export function looseToNumber(val: any): any {
  const n = Number.parseFloat(val)
  return Number.isNaN(n) ? val : n
}

export function compare<T>(value?: T, currentValue?: T, comparator?: string | ((a: T, b: T) => boolean)) {
  if (value === undefined || currentValue === undefined) {
    return false
  }

  if (typeof value === 'string') {
    return value === currentValue
  }

  if (typeof comparator === 'function') {
    return comparator(value, currentValue)
  }

  if (typeof comparator === 'string') {
    return get(value!, comparator) === get(currentValue!, comparator)
  }

  return isEqual(value, currentValue)
}

export function isEmpty(value: unknown): boolean {
  if (value == null) {
    return true
  }

  if (typeof value === 'boolean' || typeof value === 'number') {
    return false
  }

  if (typeof value === 'string') {
    return value.trim().length === 0
  }

  if (Array.isArray(value)) {
    return value.length === 0
  }

  if (value instanceof Map || value instanceof Set) {
    return value.size === 0
  }

  if (value instanceof Date || value instanceof RegExp || typeof value === 'function') {
    return false
  }

  if (typeof value === 'object') {
    for (const _ in value as object) {
      if (Object.prototype.hasOwnProperty.call(value, _)) {
        return false
      }
    }
    return true
  }

  return false
}

export function getDisplayValue<T, V>(
  items: T[],
  value: V | undefined | null,
  options: {
    valueKey?: GetItemKeys<T>
    labelKey?: keyof NestedItem<T>
  } = {}
): string | undefined {
  const { valueKey, labelKey } = options

  if (isEmpty(value)) {
    return undefined
  }

  const foundItem = items.find((item) => {
    const itemValue = (typeof item === 'object' && item !== null && valueKey)
      ? get(item, valueKey as string)
      : item
    return compare(itemValue, value)
  })

  const source = foundItem ?? value

  if (source === null || source === undefined) {
    return undefined
  }

  if (typeof source === 'object') {
    return labelKey ? get(source as Record<string, any>, labelKey as string) : undefined
  }

  return String(source)
}

export function isArrayOfArray<A>(item: A[] | A[][]): item is A[][] {
  return Array.isArray(item[0])
}

export function mergeClasses(appConfigClass?: string | string[], propClass?: string) {
  if (!appConfigClass && !propClass) {
    return ''
  }

  return [
    ...(Array.isArray(appConfigClass) ? appConfigClass : [appConfigClass]),
    propClass
  ].filter(Boolean)
}

export function getSlotChildrenText(children: any) {
  return children.map((node: any) => {
    if (!node.children || typeof node.children === 'string') return node.children || ''
    else if (Array.isArray(node.children)) return getSlotChildrenText(node.children)
    else if (node.children.default) return getSlotChildrenText(node.children.default())
  }).join('')
}

export function transformUI(ui: any, uiProp?: any) {
  return Object.entries(ui).reduce((acc, [key, value]) => {
    acc[key] = typeof value === 'function' ? value({ class: uiProp?.[key] }) : value
    return acc
  }, uiProp || {})
}

export * from './content'
