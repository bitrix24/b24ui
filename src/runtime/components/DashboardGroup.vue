<script lang="ts">
import type { AppConfig } from '@nuxt/schema'
import theme from '#build/b24ui/dashboard-group'
import type { UseResizableProps } from '../composables/useResizable'
import type { ComponentConfig } from '../types/tv'

type DashboardGroup = ComponentConfig<typeof theme, AppConfig, 'dashboardGroup'>

export interface DashboardGroupProps extends Pick<UseResizableProps, 'storage' | 'storageKey' | 'persistent' | 'unit'> {
  /**
   * The element or component this component should render as.
   * @defaultValue 'div'
   */
  as?: any
  class?: any
}

export interface DashboardGroupSlots {
  default(props?: {}): any
}
</script>

<script setup lang="ts">
import { ref, computed } from 'vue'
import { Primitive } from 'reka-ui'
import { useNuxtApp, useAppConfig } from '#imports'
import { provideDashboardContext } from '../utils/dashboard'
import { tv } from '../utils/tv'

const props = withDefaults(defineProps<DashboardGroupProps>(), {
  storage: 'cookie',
  storageKey: 'dashboard',
  persistent: true,
  unit: '%'
})
defineSlots<DashboardGroupSlots>()

const nuxtApp = useNuxtApp()
const appConfig = useAppConfig() as DashboardGroup['AppConfig']

const b24ui = computed(() => tv({ extend: tv(theme), ...(appConfig.b24ui?.dashboardGroup || {}) }))

const sidebarOpen = ref(false)
const sidebarCollapsed = ref(false)

provideDashboardContext({
  storage: props.storage,
  storageKey: props.storageKey,
  persistent: props.persistent,
  unit: props.unit,
  sidebarOpen,
  toggleSidebar: () => {
    nuxtApp.hooks.callHook('dashboard:sidebar:toggle')
  },
  sidebarCollapsed,
  collapseSidebar: (collapsed: boolean) => {
    nuxtApp.hooks.callHook('dashboard:sidebar:collapse', collapsed)
  },
  toggleSearch: () => {
    nuxtApp.hooks.callHook('dashboard:search:toggle')
  }
})
</script>

<template>
  <Primitive :as="as" :class="b24ui({ class: props.class })">
    <slot />
  </Primitive>
</template>
