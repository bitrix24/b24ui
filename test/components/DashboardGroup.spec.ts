import { describe, it, expect } from 'vitest'
import { axe } from 'vitest-axe'
import { mountSuspended } from '@nuxt/test-utils/runtime'
import DashboardGroup from '../../src/runtime/components/DashboardGroup.vue'
import type { DashboardGroupProps, DashboardGroupSlots } from '../../src/runtime/components/DashboardGroup.vue'
import ComponentRender from '../component-render'

describe('DashboardGroup', () => {
  it.each([
    // Props
    ['with as', { props: { as: 'section' } }],
    ['with class', { props: { class: 'inset-4' } }],
    // Slots
    ['with default slot', { slots: { default: () => 'Default slot' } }]
  ])('renders %s correctly', async (nameOrHtml: string, options: { props?: DashboardGroupProps, slots?: Partial<DashboardGroupSlots> }) => {
    const html = await ComponentRender(nameOrHtml, options, DashboardGroup)
    expect(html).toMatchSnapshot()
  })

  it('passes accessibility tests', async () => {
    const wrapper = await mountSuspended(DashboardGroup)

    expect(await axe(wrapper.element)).toHaveNoViolations()
  })
})
