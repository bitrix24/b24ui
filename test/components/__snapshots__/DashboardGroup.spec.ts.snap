// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DashboardGroup > renders with as correctly 1`] = `"<section class="fixed inset-0 flex overflow-hidden"></section>"`;

exports[`DashboardGroup > renders with class correctly 1`] = `"<div class="fixed flex overflow-hidden inset-4"></div>"`;

exports[`DashboardGroup > renders with default slot correctly 1`] = `"<div class="fixed inset-0 flex overflow-hidden">Default slot</div>"`;
